internal/notification/pubsub/pubsub.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/run/v1"
+)
+
+// Client represents a client to Google Cloud Pub/Sub.
+type Client interface {
+	Publish(ctx context.Context, event RolloutEvent) error
+}
+
+// PubSub is a Google Cloud Pub/Sub client to publish messages.
+type PubSub struct {
+	topic *cloudpubsub.Topic
+}
+
+// RolloutEvent is the format of an event published to Pub/Sub.
+type RolloutEvent struct {
+	Event                        string       `json:"event"`
+	CandidateRevisionName        string       `json:"candidateRevisionName"`
+	CandidateRevisionPercent     int          `json:"candidateRevisionPercent"`
+	CandidateRevisionURL         string       `json:"candidateRevisionURL"`
+	CandidateWasPromotedToStable bool         `json:"candidateWasPromotedToStable"`
+	Service                      *run.Service `json:"service"`
+}
+
+// New initializes a PubSub client to a topic in a project.
+func New(ctx context.Context, projectID string, topicName string) (ps PubSub, err error) {
+	logger := util.LoggerFrom(ctx)
+	client, err := cloudpubsub.NewClient(ctx, projectID)
+	if err != nil {
+		return ps, errors.Wrap(err, "failed to initialize Pub/Sub client")
+	}
+
+	match := regexp.MustCompile(`projects/([^/]*)/topics/([^/]*)`).FindStringSubmatch(topicName)
+	if len(match) != 3 {
+		return ps, errors.Errorf("invalid topic name %s", topicName)
+	}
+	project := match[1]
+	topicID := match[2]
+	logger.WithFields(logrus.Fields{"topicProject": project, "topicID": topicID}).Debug("parsed topic name values")
+
+	return PubSub{
+		topic: client.TopicInProject(topicID, project),
+	}, nil
+}
+
+// NewRolloutEvent initializes an event to publish to PubSub.
+func NewRolloutEvent(svc *run.Service, diagnosis health.DiagnosisResult, candidateWasPromoted bool) (RolloutEvent, error) {
+	event := "rollout"
+	if diagnosis == health.Unhealthy {
+		event = "rollback"
+	}
+
+	var candidateRevision *run.TrafficTarget
+	var err error
+	if candidateWasPromoted {
+		// The last candidate is now the stable.
+		candidateRevision, err = findRevisionWithTag(svc, "stable")
+	} else {
+		candidateRevision, err = findRevisionWithTag(svc, "candidate")
+	}
+	if err != nil {
+		return RolloutEvent{}, errors.New("failed to find candidate revision traffic target")
+	}
+
+	return RolloutEvent{
+		Event:                        event,
+		CandidateRevisionName:        candidateRevision.RevisionName,
+		CandidateRevisionPercent:     int(candidateRevision.Percent),
+		CandidateRevisionURL:         candidateRevision.Url,
+		CandidateWasPromotedToStable: candidateWasPromoted,
+		Service:                      svc,
+	}, nil
+}
+
+// Publish publishes message to the topic.
+func (ps PubSub) Publish(ctx context.Context, event RolloutEvent) error {
+	data, err := json.Marshal(event)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal message")
+	}
+
+	logger := util.LoggerFrom(ctx)
+	ps.topic.Publish(ctx, &cloudpubsub.Message{
```

Comment:
handle err

Reply:
As discussed, no errors are returned in [`Publish`](https://pkg.go.dev/cloud.google.com/go/pubsub?tab=doc#Topic.Publish) unless we call [`Get`](https://pkg.go.dev/cloud.google.com/go/pubsub?tab=doc#PublishResult.Get) on the result (blocking operation). If we're expecting `Stop` to be called at the end, calling `Get` seems unnecessary.